pkg/cli/cvd.go

```diff
@@ -278,18 +278,103 @@ func (c *cvdCreator) createCVDFromAndroidCI() ([]*hoapi.CVD, error) {
 }
 
 func (c *cvdCreator) createWithCanonicalConfig() ([]*hoapi.CVD, error) {
+	hostSrv := c.client.HostService(c.opts.Host)
+	envConfig := make(map[string]any)
+	deepCopy(c.opts.EnvConfig, &envConfig)
+	if err := c.uploadFilesFromEnvConfig(hostSrv, envConfig); err != nil {
+		return nil, fmt.Errorf("failed uploading files from environment config: %w", err)
+	}
 	createReq := &hoapi.CreateCVDRequest{
-		EnvConfig: c.opts.EnvConfig,
+		EnvConfig: envConfig,
 	}
 	c.statePrinter.Print(stateMsgFetchAndStart)
-	res, err := c.client.HostService(c.opts.Host).CreateCVD(createReq, c.credentialsFactory())
+	res, err := hostSrv.CreateCVD(createReq, c.credentialsFactory())
 	c.statePrinter.PrintDone(stateMsgFetchAndStart, err)
 	if err != nil {
 		return nil, err
 	}
 	return res.CVDs, nil
 }
 
+func (c *cvdCreator) uploadFilesFromEnvConfig(hostSrv hoclient.HostOrchestratorService, config map[string]interface{}) error {
+	if err := c.uploadCVDHostPackageFromEnvConfig(hostSrv, config); err != nil {
+		return err
+	}
+	return c.uploadImagesFromEnvConfig(hostSrv, config)
+}
+
+func (c *cvdCreator) uploadCVDHostPackageFromEnvConfig(hostSrv hoclient.HostOrchestratorService, config map[string]interface{}) error {
+	common, ok := config["common"]
```

Comment:
There is a type-safe definition of what a config file contains (using protobuffer) in the android-cuttlefish project. Consider using that instead of accessing the JSON file through map keys.

Reply:
Good point, created b/378123925 and added relevant TODOs in the code to work with the protobuf definition.  However, I'd like to do this in a follow up PR as this would require some work in the `android-cuttlefish` repo as well to make the proto available as a Golang library.

Reply:
Sounds good